components/codec/src/bits/buffered_read.rs

```diff
@@ -0,0 +1,224 @@
+// Copyright 2023 The CeresDB Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::bits::{Bit, Error};
+
+/// BufferedReader
+///
+/// BufferedReader encapsulates a buffer of bytes which can be read from.
+#[derive(Debug)]
+pub struct BufferedReader<'a> {
+    bytes: &'a [u8], // internal buffer of bytes
+    index: usize,    // index into bytes
+    pos: u32,        // position in the byte we are currently reading
+}
+
+impl<'a> BufferedReader<'a> {
+    /// new creates a new `BufferedReader` from `bytes`
+    pub fn new(bytes: &'a [u8]) -> Self {
+        BufferedReader {
+            bytes,
+            index: 0,
+            pos: 0,
+        }
+    }
+
+    fn get_byte(&mut self) -> Result<u8, Error> {
+        self.bytes.get(self.index).cloned().ok_or(Error::Eof)
+    }
+}
+
+impl<'a> BufferedReader<'a> {
+    pub fn read_bit(&mut self) -> Result<Bit, Error> {
+        if self.pos == 8 {
+            self.index += 1;
+            self.pos = 0;
+        }
+
+        let byte = self.get_byte()?;
+
+        let bit = if byte & 1u8.wrapping_shl(7 - self.pos) == 0 {
+            Bit::Zero
+        } else {
+            Bit::One
+        };
+
+        self.pos += 1;
+
+        Ok(bit)
+    }
+
+    pub fn read_byte(&mut self) -> Result<u8, Error> {
+        if self.pos == 0 {
+            self.pos += 8;
+            return self.get_byte();
+        }
+
+        if self.pos == 8 {
+            self.index += 1;
+            return self.get_byte();
+        }
+
+        let mut byte = 0;
+        let mut b = self.get_byte()?;
+
+        byte |= b.wrapping_shl(self.pos);
+
+        self.index += 1;
+        b = self.get_byte()?;
+
+        byte |= b.wrapping_shr(8 - self.pos);
+
+        Ok(byte)
+    }
+
+    // example: read_bits(4): read data:u64 0000 0000 0000 000f
+    // datastore in low position
```

Comment:
Use the rust doc comments, that is to say, use `///` instead of `//`.

And the comment here are too casual and informal, shall we polish it up and make it more formal?